sys: Fix warning for stripped const qualifier

yperess opened the pull request:
Fix several errors caused by adding stricter warnings used by pigweed

Signed-off-by: Yuval Peress <peress@google.com>

Commit:
sys: Fix warning for stripped const qualifier

This causes an issue when compiling with `-Werror=cast-qual`

Signed-off-by: Yuval Peress <peress@google.com>

Commit:
cbprintf: Fix unused arg warning

This was found when building with `-Wextra` enabled.

Signed-off-by: Yuval Peress <peress@google.com>

Commit:
assert: Add missing include

printk is used if CONFIG_ASSERT_VERBOSE is set but was missing an import

Signed-off-by: Yuval Peress <peress@google.com>

jackrosenthal commented on include/sys/cbprintf_internal.h:
```diff
@@ -314,7 +314,7 @@ union z_cbprintf_hdr {
  * the macro which generates the warning to not slow down the compiler.
  */
-#if __clang__ == 1
+#if defined(__clang__) && __clang__ == 1
```
`#ifdef __clang__`

yperess replied:
Done

jackrosenthal commented on include/logging/log_msg2.h:
```diff
@@ -28,7 +28,7 @@ extern "C" {
 #define LOG_MSG2_DBG(...) IF_ENABLED(LOG_MSG2_DEBUG, (printk(__VA_ARGS__)))
 
-#if CONFIG_LOG_TIMESTAMP_64BIT
+#ifdef CONFIG_LOG_TIMESTAMP_64BIT
```
these unrelated changes seem fine, but are they necessary for the missing const?

nordic-krch replied:
yes, you could split those into separate commit.

yperess replied:
Done

jackrosenthal commented on include/sys/cbprintf_cxx.h:
```diff
@@ -52,6 +52,7 @@ static inline int z_cbprintf_cxx_is_pchar(const volatile wchar_t *)
 template < typename T >
 static inline int z_cbprintf_cxx_is_pchar(T arg)
 {
+	ARG_UNUSED(arg);
```
same as this, seems unrelated to the missing const (should be separate PR)?

nordic-krch replied:
I think that it's fine as long as it's in separate commit.

yperess replied:
This is in its own commit (https://github.com/zephyrproject-rtos/zephyr/pull/39932/commits/018b440cc9e950c61634578036f46e6358227a6f)

Commit:
sys: cbprintf: Fix unused arg warning

Missed unused arg, throws warning when compiling with
-Wunused-parameter.

Signed-off-by: Yuval Peress <peress@google.com>

Commit:
fatal: fix missing extern "C"

This was causing an error when building with pigweed as a module.

Signed-off-by: Yuval Peress <peress@google.com>

Commit:
zephyr: fix errors when building with -Wundef

Several points check the value of CONFIG_s without first checking that
they're defined. This causes an warning when adding -Wundef to the
build.

Signed-off-by: Yuval Peress <peress@google.com>

MaureenHelm reviewed (commented):
Need to fix the failed compliance check:
```
Error: See docs.zephyrproject.org/latest/contribute/#commit-guidelines for more details

Commit d67b18498a:
1: UC6 Body has no content, should at least have 1 line.

Commit 40f5517568:
1: UC6 Body has no content, should at least have 1 line.
```

ElectricalPaul commented:
Sorry about the approval spam. I reviewed each individual commit in the stack and approved them one at a time. Apparently, approving applies to the entire stack, not to an individual commit. I guess I need to learn Github's system a little better ...

yperess commented:
> Need to fix the failed compliance check:
> 
> ```
> Error: See docs.zephyrproject.org/latest/contribute/#commit-guidelines for more details
> 
> Commit d67b18498a:
> 1: UC6 Body has no content, should at least have 1 line.
> 
> Commit 40f5517568:
> 1: UC6 Body has no content, should at least have 1 line.
> ```

Done

nordic-krch commented on include/sys/cbprintf_cxx.h:
```diff
@@ -57,6 +58,7 @@ static inline int z_cbprintf_cxx_is_pchar(const volatile wchar_t *, bool const_a
 static inline int z_cbprintf_cxx_is_pchar(T arg, bool const_as_fixed)
 {
+	ARG_UNUSED(arg);
 	_Pragma("GCC diagnostic push")
```
can you put ARG_UNUSEDs next to each other? You can move `ARG_UNUSED(const_as_fixed);` up as well.

nordic-krch reviewed (approved):
looks ok, just one minor comment
